[WASM] Compute B3 effect ranges for memory and global ops in OMG

ddegazio opened the pull request:
#### 958d1de8af7b3bea8e5df94bcf029935abc7b3e0
<pre>
[WASM] Compute B3 effect ranges for memory and global ops in OMG
<a href="https://bugs.webkit.org/show_bug.cgi?id=276504">https://bugs.webkit.org/show_bug.cgi?id=276504</a>
<a href="https://rdar.apple.com/problem/131562611">rdar://problem/131562611</a>

Reviewed by Keith Miller, Justin Michaud and Yusuke Suzuki.

Adds an AbstractHeap notion to the OMG tier to represent ranges of effects. We
compute distinct HeapRanges associated with each of these heaps after compilation,
and pass them along to B3 in order to help with CSE and potentially hoisting.
This initial patch defines abstract heap kinds for globals, WASM memory, and the
fields of a variety of WASM structures such as the instance or tables. We should
consider adding more later for things like GC types or the JS API, and/or introduce
a parent/child system for AbstractHeaps like we have in FTL.

* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
(JSC::Wasm::OMGIRGenerator::AbstractHeap::AbstractHeap):
(JSC::Wasm::OMGIRGenerator::recordHeap):
(JSC::Wasm::OMGIRGenerator::computeHeapRanges):
(JSC::Wasm::OMGIRGenerator::emitPrepareWasmOperation):
(JSC::Wasm::OMGIRGenerator::insertConstants):
(JSC::Wasm::OMGIRGenerator::addCurrentMemory):
(JSC::Wasm::OMGIRGenerator::getGlobal):
(JSC::Wasm::OMGIRGenerator::setGlobal):
(JSC::Wasm::OMGIRGenerator::emitWriteBarrierForJSWrapper):
(JSC::Wasm::OMGIRGenerator::emitLoadOp):
(JSC::Wasm::OMGIRGenerator::load):
(JSC::Wasm::OMGIRGenerator::emitStoreOp):
(JSC::Wasm::OMGIRGenerator::emitAtomicLoadOp):
(JSC::Wasm::OMGIRGenerator::emitAtomicStoreOp):
(JSC::Wasm::OMGIRGenerator::emitAtomicBinaryRMWOp):
(JSC::Wasm::OMGIRGenerator::emitAtomicCompareExchange):
(JSC::Wasm::OMGIRGenerator::addSIMDLoad):
(JSC::Wasm::OMGIRGenerator::addSIMDStore):
(JSC::Wasm::OMGIRGenerator::addSIMDLoadSplat):
(JSC::Wasm::OMGIRGenerator::addSIMDLoadLane):
(JSC::Wasm::OMGIRGenerator::addSIMDStoreLane):
(JSC::Wasm::OMGIRGenerator::addSIMDLoadExtend):
(JSC::Wasm::OMGIRGenerator::addSIMDLoadPad):
(JSC::Wasm::OMGIRGenerator::emitInlineDirectCall):
(JSC::Wasm::OMGIRGenerator::addCall):
(JSC::Wasm::OMGIRGenerator::addCallIndirect):
(JSC::Wasm::parseAndCompileOMG):

Canonical link: <a href="https://commits.webkit.org/281143@main">https://commits.webkit.org/281143@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/80fde9e7aee78212e3ac40285ce3d2dd2f70a09b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58808 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38136 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11298 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62440 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9252 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45773 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9452 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47570 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6585 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60839 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35678 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50850 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28426 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32412 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8141 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8256 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/51901 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54367 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8421 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/64141 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/58050 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2721 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8407 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54892 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2730 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50877 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54984 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13035 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2276 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/79811 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/33966 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/13803 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/35050 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36135 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34796 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud reviewed (approved):
r=me

Nice. It would be nicer to have it integrated with the FTL version, for the future where we can inline wasm->js

Achierius commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -750,12 +750,82 @@ class OMGIRGenerator {
+        Kind kind;
+        unsigned offset : 24;
+
+        inline AbstractHeap(Kind kind_in)
```
Does the `inline` specifier do anything here? I thought it was implicit for all member functions.

kmiller68 replied:
I don't think it does. ALWAYS_INLINE would do something. `inline` also does something when the definition is out of line (It errors if the definition isn't defined in the translation unit rather than a link error).

ddegazio replied:
Just a reflex since I'm used to defining these types of types in a header. I can remove it! 😅

ddegazio replied:
Resolved in latest push, actually some rewrites mean this `Kind` is no longer needed.

Achierius commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -750,12 +750,82 @@ class OMGIRGenerator {
 #if !USE(BUILTIN_FRAME_ADDRESS) || ASSERT_ENABLED
         // Prepare wasm operation calls.
-        block->appendNew<B3::MemoryValue>(m_proc, B3::Store, origin(), framePointer(), instanceValue(), Instance::offsetOfTemporaryCallFrame());
+        recordHeap(block->appendNew<B3::MemoryValue>(m_proc, B3::Store, origin(), framePointer(), instanceValue(), Instance::offsetOfTemporaryCallFrame()), AbstractHeap::InstanceFields);
```
Why does this depend on whether or not asserts are enabled? Is this feature debug-only?

justinmichaud replied:
This is for debugging the code that finds the callframe inside DECLARE_WASM_CALL_FRAME, see https://commits.webkit.org/258820@main

kmiller68 commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -750,12 +750,82 @@ class OMGIRGenerator {
+        };
+
+        Kind kind;
+        unsigned offset : 24;
```
Does this actually make a bitfield? Technically, I thought all members have to be the same type. Although maybe that's changed in newer C++ versions...

Achierius replied:
I was wondering about something similar and plugged it into compiler explorer, the sizeof the whole structure does come out to 4B so it looks to be working as expected.

ddegazio replied:
Good catch...that's weird that it came out as 4B, but I'll define `Kind : unsigned` to guarantee it.

Achierius commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -750,12 +750,82 @@ class OMGIRGenerator {
+        std::array<unsigned, AbstractHeap::NumKinds> startIndexPerKind;
+        startIndexPerKind[0] = 0;
+        for (unsigned i = 0; i < AbstractHeap::NumKinds - 1; i ++)
+            startIndexPerKind[i + 1] = startIndexPerKind[i] + maxIndicesPerKind[i];
```
I believe you can combine this scan op into the prior loop; I don't know how big the array is so it might not matter, but I don't think the compiler will fuse them on its own.

ddegazio replied:
Refactored this function in the latest push, so this is no longer relevant.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -1756,7 +1828,7 @@ auto OMGIRGenerator::addCurrentMemory(ExpressionType& result) -> PartialResult
     static_assert(sizeof(std::declval<Memory*>()->size()) == sizeof(uint64_t), "codegen relies on this size");
 
-    Value* memory = m_currentBlock->appendNew<MemoryValue>(m_proc, Load, Int64, origin(), instanceValue(), safeCast<int32_t>(Instance::offsetOfMemory()));
+    Value* memory = recordHeap(m_currentBlock->appendNew<MemoryValue>(m_proc, Load, Int64, origin(), instanceValue(), safeCast<int32_t>(Instance::offsetOfMemory())), AbstractHeap::InstanceFields);
```
It's a little unfortunate that we don't have an FTLOutput equivalent here since we could've forced a heap if we did. I have a feeling someone is going to forget to add an AbstractHeap at some point and it's going to cause problems... Maybe MemoryValue should require a HeapRange in the constructor or we should ASSERT no MemoryValue has a top range?

ddegazio replied:
I think adding it to the MemoryValue constructor would only help a little, since the easier thing to forget is actually adding the value to the list of memory values so we can compute the index ranges later. Maybe we can make an `OMGMemoryValue` that takes a reference to the generator and adds itself to the list?

ddegazio replied:
Changed a lot of the patch to facilitate this - we now use an `append<ValueType>(...)` function in OMG to add values, not `block->appendNew` directly. With this new API point we can define adapters that require additional parameters (like an abstract heap) based on the value type (for instance `MemoryValue`). I've applied this to all memory values in OMG and hopefully it'll be easier to remember this way in the future (since as long as you're using `OMGIRGenerator::append`, you'll get a compiler error if you try and add a memory value without a specified heap).

kmiller68 reviewed (approved):
r=me with questions.

Constellation reviewed (approved):
r=me

ddegazio committed:
[WASM] Compute B3 effect ranges for memory and global ops in OMG
https://bugs.webkit.org/show_bug.cgi?id=276504
rdar://problem/131562611

Reviewed by Keith Miller, Justin Michaud and Yusuke Suzuki.

Adds an AbstractHeap notion to the OMG tier to represent ranges of effects. We
compute distinct HeapRanges associated with each of these heaps after compilation,
and pass them along to B3 in order to help with CSE and potentially hoisting.
This initial patch defines abstract heap kinds for globals, WASM memory, and the
fields of a variety of WASM structures such as the instance or tables. We should
consider adding more later for things like GC types or the JS API, and/or introduce
a parent/child system for AbstractHeaps like we have in FTL.

* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
(JSC::Wasm::OMGIRGenerator::AbstractHeap::AbstractHeap):
(JSC::Wasm::OMGIRGenerator::recordHeap):
(JSC::Wasm::OMGIRGenerator::computeHeapRanges):
(JSC::Wasm::OMGIRGenerator::emitPrepareWasmOperation):
(JSC::Wasm::OMGIRGenerator::insertConstants):
(JSC::Wasm::OMGIRGenerator::addCurrentMemory):
(JSC::Wasm::OMGIRGenerator::getGlobal):
(JSC::Wasm::OMGIRGenerator::setGlobal):
(JSC::Wasm::OMGIRGenerator::emitWriteBarrierForJSWrapper):
(JSC::Wasm::OMGIRGenerator::emitLoadOp):
(JSC::Wasm::OMGIRGenerator::load):
(JSC::Wasm::OMGIRGenerator::emitStoreOp):
(JSC::Wasm::OMGIRGenerator::emitAtomicLoadOp):
(JSC::Wasm::OMGIRGenerator::emitAtomicStoreOp):
(JSC::Wasm::OMGIRGenerator::emitAtomicBinaryRMWOp):
(JSC::Wasm::OMGIRGenerator::emitAtomicCompareExchange):
(JSC::Wasm::OMGIRGenerator::addSIMDLoad):
(JSC::Wasm::OMGIRGenerator::addSIMDStore):
(JSC::Wasm::OMGIRGenerator::addSIMDLoadSplat):
(JSC::Wasm::OMGIRGenerator::addSIMDLoadLane):
(JSC::Wasm::OMGIRGenerator::addSIMDStoreLane):
(JSC::Wasm::OMGIRGenerator::addSIMDLoadExtend):
(JSC::Wasm::OMGIRGenerator::addSIMDLoadPad):
(JSC::Wasm::OMGIRGenerator::emitInlineDirectCall):
(JSC::Wasm::OMGIRGenerator::addCall):
(JSC::Wasm::OMGIRGenerator::addCallIndirect):
(JSC::Wasm::parseAndCompileOMG):

Canonical link: https://commits.webkit.org/281143@main
